spec/controllers/receipts_controller_spec.rb

```diff
@@ -0,0 +1,67 @@
+require "rails_helper"
+
+RSpec.describe ReceiptsController, type: :controller do
+  let!(:user) { FactoryBot.create :user }
+  let!(:receipt) { FactoryBot.create :receipt, user_id: user.id }
+  let!(:receipts) { user.receipts }
+
+  before do
+    log_in user
+  end
+
+  describe "GET receipts#index" do
+    context "success when valid attributes" do
+      before do
+        get :index
+      end
+
+      it "assigns @receipts" do
+        expect(assigns(:receipts)).to eq receipts
+      end
+
+      it {should render_template(:index)}
+    end
+  end
+
+  describe "GET receipts#show" do   
+    context "success when valid attributes" do
+      before do
+        get :show, params: { id: receipt.id }
+      end
+
+      it "assigns @receipt exist" do
+        expect(assigns(:receipt)).to eq receipt
+      end
+
+      it {should render_template(:show)}
+    end
+
+    context "success when receipt not exist" do
+      before do
+        get :show, params: { id: -1 }
+      end
+
+      it "show flash danger" do
+        expect(flash[:danger]).to eq I18n.t("receipt.not_permissions")
+      end
+
+      it {should redirect_to(receipts_path)}
+    end
+  end
+
+  describe "POST receipts#cancel_booking" do
+    context "success when valid attributes" do
```

Comment:
test cả việc sau khi gọi xong action cancel thì cái receip đó sẽ dc thay đổi status thành cancel nhé

ngoài ra, test việc nếu receipt waiting, thì test cả trường hợp ko phải wait thì status sẽ ko dc change nhé

Reply:
như vậy em sẽ test 2 context gồm 
1. when valid attributes: status receipt is wait and user logged
   cái này mong đợi sẽ check: flash, render/redirect_to, change data ?...
2. when invalid attributes
   2.1. receipt is not found
   2.2 status receipt is not wait
   2.3 user not login